File: extensions/husky-analyzer/src/data/Focus.ts
import { decode_member, decode_number, decode_opt } from "src/decode/decode";
import type Trace from "src/trace/Trace";

export default class Focus {
    opt_input_id: number | null;

    constructor(data?: unknown) {
        if (data === undefined) {
            this.opt_input_id = null;
        } else {
            this.opt_input_id = decode_opt(
                decode_member(data, "opt_input_id"),
                decode_number
            );
        }
    }

    gen_figure_key(trace_id: number): string {
        return `${this.key()}:${trace_id}`;
    }

    key(): string {
        if (this.opt_input_id === null) {
            return "g";
        } else {
            return JSON.stringify(this.opt_input_id);
        }
    }

    gen_subtraces_key(trace: Trace): string {
        return `${this.gen_subtraces_effective_opt_input_id(trace)}:${
            trace.id
        }`;
    }

    gen_subtraces_effective_opt_input_id(trace: Trace): number | null {
        switch (trace.kind) {
            case "Main":
            case "FeatureStmt":
            case "FuncStmt":
            case "ProcStmt":
            case "LoopFrame":
            case "FeatureBranch":
            case "FeatureCallInput":
                return null;
            case "FeatureExpr":
            case "CallHead":
            case "EagerExpr":
                return this.opt_input_id;
        }
    }
}

File: core/crates/trace/husky-tracer-protocol/src/focus.rs
use super::*;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq, Eq, Hash)]
pub enum Focus {
    Specific { input_id: usize },
    Generic {},
}

impl Default for Focus {
    fn default() -> Self {
        Focus::Generic {}
    }
}

impl Focus {
    pub fn has_stalk(&self, trace_kind: TraceKind) -> bool {
        match trace_kind {
            TraceKind::Main
            | TraceKind::FeatureStmt
            | TraceKind::FeatureBranch
            | TraceKind::FeatureExpr => match self {
                Focus::Specific { .. } => true,
                Focus::Generic {} => false,
            },
            TraceKind::FeatureCallInput
            | TraceKind::FuncStmt
            | TraceKind::ProcStmt
            | TraceKind::ProcBranch
            | TraceKind::LoopFrame
            | TraceKind::EagerExpr
            | TraceKind::CallHead => false,
        }
    }

    pub fn opt_input_id(&self) -> Option<usize> {
        match self {
            Focus::Specific { input_id } => Some(*input_id),
            Focus::Generic {} => None,
        }
    }
}
// function tell_has_extra(trace: Trace, input_id: number | null): boolean {
//     switch (trace.kind) {
//         case "Main":
//         case "FeatureStmt":
//         case "FeatureBranch":
//         case "FeatureExpr":
//             return input_id !== null;
//         case "FuncStmt":
//         case "LoopFrame":
//         case "ProcStmt":
//         case "EagerExpr":
//         case "CallHead":
//         case "FeatureCallInput":
//         case "LoopFrame":
//         case "ProcBranch":
//             return false;
//     }
// }
